### PR TITLE
Migration from Spring RestTemplate to OkHttp

### PR DESCRIPTION
**Pull request by pellepelster**
Summary:

* all RestTemplate calls are now replaced with OkHttp equivalents, and Spring dependencies remove
* extended Tests for ssh key, certificate and loadbalancer management
* Minor documentation updates

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -95,33 +88,26 @@
...
-                <exclusion>
-                    <groupId>junit</groupId>
-                    <artifactId>junit</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
         <dependency>
             <groupId>org.junit.jupiter</groupId>
-            <artifactId>junit-jupiter-engine</artifactId>
-            <scope>test</scope>
```
Comment (tomsiewert) [inline pom.xml:112]: Please add the test scope back

File: src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java [hunk 2]
```diff
@@ -1,56 +1,47 @@
...
-    private final HttpEntity<String> httpEntity;
-    private final HttpHeaders httpHeaders;
-    private final RestTemplate restTemplate;
+    private final String hcloudToken;
+    private final ObjectMapper objectMapper;
 
     /**
      * Initial method to use the API
      *
      * @param token API-Token for Hetzner Cloud API
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java:31]: Rename to `hcloudToken` or change the var back to `token`.

File: src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java [hunk 3]
```diff
@@ -1,56 +1,47 @@
...
      */
-    public HetznerCloudAPI(String token) {
-        this.token = token;
+    public HetznerCloudAPI(String hcloudToken) {
 
-        List<HttpMessageConverter<?>> messageConverters = new ArrayList<>();
-        MappingJackson2HttpMessageConverter converter = new MappingJackson2HttpMessageConverter();
-        converter.setSupportedMediaTypes(Collections.singletonList(MediaType.APPLICATION_JSON));
-        messageConverters.add(converter);
+        if (hcloudToken == null || hcloudToken.isBlank()) {
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java:36]: Unnecessary brackets.

File: src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java [hunk 4]
```diff
@@ -2381,22 +2190,62 @@ public PlacementGroupResponse createPlacementGroup(PlacementGroupRequest placeme
...
+                String.class);
     }
 
-    /**
-     * Converts a Date to the ISO-8601 format
-     *
-     * @param date Date to be converted
-     * @return Date in ISO-8601 format
-     */
-    public String convertToISO8601(Date date) {
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java:2397]: Is there any reason why you've removed this method? I would more like to deprecate it first and remove it with a next release if you think it shouldn't be here.

File: src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java [hunk 5]
```diff
@@ -2381,22 +2190,62 @@ public PlacementGroupResponse createPlacementGroup(PlacementGroupRequest placeme
...
+
+    private <T> T post(String url, Object body, Class<T> clazz) {
+        return exchange(url, HttpMethod.POST, body, clazz);
+    }
+
+    private <T> T post(String url, Class<T> clazz) {
+        return exchange(url, HttpMethod.POST, null, clazz);
+    }
+
+    public enum HttpMethod {
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java:2247]: Do we really need this class being public?

File: src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class UrlBuilder {
+
+    private String url;
+
+    private Map<String, String> queryParams = new HashMap<>();
+
+    private UrlBuilder(String url) {
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java:15]: We can generate this via Lombok

File: src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        this.url = url;
+    }
+
+    public static UrlBuilder from(String url) {
+
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java:20]: This class has a lot of unnecessary empty newlines. Please remove them.

File: src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        if (url == null || url.isBlank()) {
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java:21]: Unnecessary brackets.

File: src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+            throw new RuntimeException("url must not be empty or blank");
+        }
+
+        return new UrlBuilder(url);
+    }
+
+    public UrlBuilder queryParamIfPresent(String name, Optional<Integer> value) {
+        value.ifPresent((it) -> queryParams.put(name, it.toString()));
+        return this;
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java:30]: If I'm not wrong you don't need to do `return this` and return the `value.ifPresent` directly.

File: src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
+package me.tomsdevsn.hetznercloud;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class UrlBuilder {
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java:9]: Please add javadocs to this class

File: src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerChangeAlgorithmRequest.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+@AllArgsConstructor
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerChangeAlgorithmRequest.java:12]: Please remove this as it only includes properties which are all required.

File: src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerChangeProtectionRequest.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+@AllArgsConstructor
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerChangeProtectionRequest.java:12]: same here

File: src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerChangeTypeRequest.java [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+@AllArgsConstructor
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerChangeTypeRequest.java:12]: same here

File: src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerDeleteServiceRequest.java [hunk 14]
```diff
@@ -0,0 +1,22 @@
+package me.tomsdevsn.hetznercloud.objects.request;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import me.tomsdevsn.hetznercloud.objects.general.LBService;
+import me.tomsdevsn.hetznercloud.objects.general.LBTarget;
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerDeleteServiceRequest.java:9]: Cleanup imports

File: src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerDeleteServiceRequest.java [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+
+import java.util.List;
+import java.util.Map;
+
+@Data
+@Builder
+@AllArgsConstructor
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerDeleteServiceRequest.java:17]: same here

File: src/main/java/me/tomsdevsn/hetznercloud/objects/request/PlacementGroupAddServerRequest.java [hunk 16]
```diff
@@ -0,0 +1,14 @@
+package me.tomsdevsn.hetznercloud.objects.request;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+@Getter
+@AllArgsConstructor
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/objects/request/PlacementGroupAddServerRequest.java:10]: same here

**Review by tomsiewert — CHANGES_REQUESTED** [review]
First of all, thank you for your great work! There are some things which need to be fixed/changed. Please see the review.

File: src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerChangeProtectionRequest.java (hunk 12 above)
Comment (tomsiewert) [reply src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerChangeProtectionRequest.java:10]: I think you accidentally removed the wrong annotation.

File: src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerDeleteServiceRequest.java [hunk 17]
```diff
@@ -1,22 +1,15 @@
...
 @Data
 @Builder
-@AllArgsConstructor
-@JsonInclude(JsonInclude.Include.NON_NULL)
+@RequiredArgsConstructor
 public class LoadBalancerDeleteServiceRequest {
 
     @JsonProperty("listen_port")
-    private long listenPort;
+    private final long listenPort;
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerDeleteServiceRequest.java:14]: I am not quite sure if we need to set it as `final`.

Also, please use the `Long` class instead of data type `long` (capitalize `L`). Using the data type made some weird issues in the past.

File: src/main/java/me/tomsdevsn/hetznercloud/objects/request/PlacementGroupAddServerRequest.java [hunk 18]
```diff
@@ -1,13 +1,11 @@
...
 import com.fasterxml.jackson.annotation.JsonProperty;
 import lombok.AllArgsConstructor;
 import lombok.Getter;
 
 @Getter
 @AllArgsConstructor
-@JsonInclude(JsonInclude.Include.NON_NULL)
 public class PlacementGroupAddServerRequest {
     @JsonProperty("placement_group")
     private long placementGroup;
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/objects/request/PlacementGroupAddServerRequest.java:11]: Please use the `Long` class instead of data type `long` (capitalize `L`). Using the data type made some weird issues in the past.

**Review by tomsiewert — CHANGES_REQUESTED** [review]

File: src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerNetworkRequest.java [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+
+@Data
+@Builder
+@AllArgsConstructor
+@RequiredArgsConstructor
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class LoadBalancerNetworkRequest {
+
+    @JsonProperty("network")
+    private final long network;
```
Comment (tomsiewert) [inline src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerNetworkRequest.java:18]: Not sure about the `final`, but I guess you need it because of `RequiredArgsConstructor` right?

Anyway, please use the `Long` class instead of data type `long` (capitalize `L`). Using the data type made some weird issues in the past.

**Review by tomsiewert — CHANGES_REQUESTED** [review]

File: src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerDeleteServiceRequest.java (hunk 14 above)
Comment (pellepelster) [reply src/main/java/me/tomsdevsn/hetznercloud/objects/request/LoadBalancerDeleteServiceRequest.java]: I added a minimal checkstyle that checks for unused imports and fixed the errors in various placed. This is PR will now be stale for a few weeks from my side due to summer vacations

File: src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java (hunk 4 above)
Comment (pellepelster) [reply src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java:2397]: It was unused, also the test method for it was empty. If its part of the public API I will re-add it.

Comment (tomsiewert) [reply src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java:2397]: The initial test suite was written by someone else, I didn't find any time to finish the tests.

And yes, it was part of the Public API. I would please you to add it back.

Comment (pellepelster) [reply src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java]: I re-added the method

File: src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java (hunk 3 above)
Comment (pellepelster) [reply src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java:40]: You mean the curly braces? Yes they are optional, but I would argue from my personal experience that the amount of bugs caused by missing braces I have seen in codebases over the years (most notably example would perhaps to a certain degree be https://www.imperialviolet.org/2014/02/22/applebug.html) by a huge margin outweighs any advantage gained by omitting those braces.

**Comment by tomsiewert** [comment]
There are still some open changes. Would you please check them?

File: src/main/java/me/tomsdevsn/hetznercloud/objects/request/PlacementGroupAddServerRequest.java (hunk 18 above)
Comment (tomsiewert) [reply src/main/java/me/tomsdevsn/hetznercloud/objects/request/PlacementGroupAddServerRequest.java:11]: I think this got over-written with a force-push :D

**Review by tomsiewert — CHANGES_REQUESTED** [review]

File: src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java (hunk 6 above)
Comment (pellepelster) [reply src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java]: constructor now comes from lombok

File: src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java (hunk 7 above)
Comment (pellepelster) [reply src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java]: should be cleaner now

File: src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java (hunk 9 above)
Comment (pellepelster) [reply src/main/java/me/tomsdevsn/hetznercloud/UrlBuilder.java:44]: return this was intentional to allow for fluent builder like construction of urls

File: src/main/java/me/tomsdevsn/hetznercloud/objects/request/PlacementGroupAddServerRequest.java (hunk 18 above)
Comment (tomsiewert) [reply src/main/java/me/tomsdevsn/hetznercloud/objects/request/PlacementGroupAddServerRequest.java:13]: Oh not just the non-null attributes, but also the data type :smile:

Comment (pellepelster) [reply src/main/java/me/tomsdevsn/hetznercloud/objects/request/PlacementGroupAddServerRequest.java]: it has been a long day, I should stop for today

**Review by tomsiewert — APPROVED** [review]
LGTM!

Thank You for your effort making this project more awesome!

It will take a bit until I make a major release (3.0.0) as I want to implement flexible networks (and firewalls) before.
